### PR TITLE
[FLINK-11963][History Server]Add time-based cleanup mechanism in history server

### PR DESCRIPTION
**Pull request by maqingxiang**
## What is the purpose of the change

The purpose of this pull request is to add time-based cleanup mechanism in history server.

### CODE REVIEW

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/history/HistoryServerArchiveFetcher.java [hunk 1]
```diff
@@ -153,9 +157,15 @@ public void run() {
 						continue;
 					}
 					boolean updateOverview = false;
+					long now = System.currentTimeMillis();
 					for (FileStatus jobArchive : jobArchives) {
 						Path jobArchivePath = jobArchive.getPath();
 						String jobID = jobArchivePath.getName();
+						if (retainedApplicationsMillis != 0L && now - jobArchive.getModificationTime() > retainedApplicationsMillis) {
```
Comment (klion26) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/history/HistoryServerArchiveFetcher.java]: Could we assert that `retainedApplicationsMillis` is always no negative?

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/history/HistoryServerArchiveFetcher.java [hunk 2]
```diff
@@ -153,9 +157,15 @@ public void run() {
...
+							LOG.debug("delete the old archived job for path {}." + jobArchivePath.toString());
```
[the remark this answers is not in the document]
Comment (maqingxiang) [reply flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/history/HistoryServerArchiveFetcher.java]: Thank you for the review, It has been modified.

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/history/HistoryServerArchiveFetcher.java (hunk 1 above)
Comment (maqingxiang) [reply flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/history/HistoryServerArchiveFetcher.java]: Thank you for the review, change the default value to -1.

Comment (klion26) [reply flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/history/HistoryServerArchiveFetcher.java]: Sorry for the ambiguous description, what I mean is "If a user sets a negative number, what will the behavior be?"
Maybe we should add a check logic

Comment (maqingxiang) [reply flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/history/HistoryServerArchiveFetcher.java]: Thank you for the review @klion26 , negative or zero value number is considered to disable the cleanup mechanism.
Add the hint statements in a log.warn fashion.

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/history/HistoryServerArchiveFetcher.java [hunk 3]
```diff
@@ -153,9 +157,20 @@ public void run() {
...
 						Path jobArchivePath = jobArchive.getPath();
 						String jobID = jobArchivePath.getName();
+						if (retainedApplicationsMillis > 0L && now - jobArchive.getModificationTime() > retainedApplicationsMillis) {
+							if (LOG.isDebugEnabled()) {
+								LOG.debug("delete the old archived job for path {}." + jobArchivePath.toString());
+							}
+							jobArchivePath.getFileSystem().delete(jobArchivePath, false);
+							continue;
+						} else {
+							LOG.warn("Negative or zero values {} of the historyserver.archive.fs.retained-application-millis " +
```
Comment (klion26) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/history/HistoryServerArchiveFetcher.java]: Maybe we will come into the else branch when `retainedApplicationsMillis` > 0?

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/history/HistoryServerArchiveFetcher.java [hunk 4]
```diff
@@ -123,7 +125,9 @@ void stop() {
 
 		private static final String JSON_FILE_ENDING = ".json";
 
-		JobArchiveFetcherTask(List<HistoryServer.RefreshLocation> refreshDirs, File webDir, CountDownLatch numFinishedPolls) {
+		JobArchiveFetcherTask(long retainedApplicationsMillis, List<HistoryServer.RefreshLocation> refreshDirs, File webDir,
+							CountDownLatch numFinishedPolls) {
+			this.retainedApplicationsMillis = retainedApplicationsMillis;
```
Comment (klion26) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/history/HistoryServerArchiveFetcher.java]: Maybe we could add a test logic whether the passed in parameter `retainedApplicationsMillis` is positive or not here

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/history/HistoryServerArchiveFetcher.java (hunk 3 above)
Comment (maqingxiang) [reply flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/history/HistoryServerArchiveFetcher.java]: I am very sorry that my carelessness caused this mistake, Thanks a lot for reviewing it again!

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/history/HistoryServerArchiveFetcher.java (hunk 4 above)
Comment (maqingxiang) [reply flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/history/HistoryServerArchiveFetcher.java]: The test logic has been added by increasing the isCleanupEnabled variable. Thanks a lot for reviewing it again!

**Comment by maqingxiang** [comment]
Looking forward to your review. @klion26

**Comment by tillrohrmann** [comment]
Thanks for creating this PR @maqingxiang and sorry that we did not manage to merge it. By now it has become a bit out dated as the `HistoryServer` code base has changed. There is in fact another similar effort which you can find here #12636. Do you want to rebase your work on this PR? I think it is pretty much complementary to each other. 

I would close this PR because it is out dated and not mergeable anymore.

**Pull request closed without merging**